Snapshotting of mainframe and subframe PDFs are blank in UI-side compositing mode

nmoucht opened the pull request:
#### 30db210558cfc9c0600560297b4b312a1afc8e3a
<pre>
Snapshotting of mainframe and subframe PDFs are blank in UI-side compositing mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=249065">https://bugs.webkit.org/show_bug.cgi?id=249065</a>
rdar://103160251

Reviewed by Simon Fraser.

Use an in process buffer when snapshotting for a frame containing a pdf. For
subframe pdfs, create a WebCore::Image using the underlying data of the
ShareableBitmap (which creates a CGImage using ShareableBitmap data as the
CGDataProviderRef) and draw that into the GraphicsContext displaylist,
which gets the image to the GPU process.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::snapshotAtSize):

Canonical link: <a href="https://commits.webkit.org/258399@main">https://commits.webkit.org/258399@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31a61e0e03b36b7779388ba9b8a6e5d3a36afded

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100697 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33740 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109999 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10783 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/352 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93101 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107839 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106479 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8145 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91389 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34754 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90055 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22782 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77727 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3538 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24310 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3561 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/356 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9673 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43805 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5342 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3037 "Failed to push commit to Webkit repository") | | | | 
<!--EWS-Status-Bubble-End-->

smfr reviewed (approved):
Can we test this?

smfr reviewed (changes requested):
Actually this makes all snapshotting happen in the web process. Let's hold off on this for now, pending discussion.

smfr commented on Source/WebKit/WebProcess/Plugins/PluginView.cpp:
```diff
@@ -428,8 +428,15 @@ void PluginView::paint(GraphicsContext& context, const IntRect& /*dirtyRect*/, W
-    if (m_transientPaintingSnapshot)
-        m_transientPaintingSnapshot->paint(context, m_plugin->deviceScaleFactor(), frameRect().location(), m_transientPaintingSnapshot->bounds());
+    if (m_transientPaintingSnapshot) {
+        if (!context.platformContext()) {
```
Do we hit this for the "update control tints" and "detecting largest contentful paint" fake paints?

nmoucht replied:
No since those pass a `NullGraphicsContext` which early returns in this function when checking `context.paintingDisabled()`.

nmoucht committed:
Snapshotting of mainframe and subframe PDFs are blank in UI-side compositing mode
https://bugs.webkit.org/show_bug.cgi?id=249065
rdar://103160251

Reviewed by Simon Fraser.

Use an in process buffer when snapshotting for a frame containing a pdf. For
subframe pdfs, create a WebCore::Image using the underlying data of the
ShareableBitmap (which creates a CGImage using ShareableBitmap data as the
CGDataProviderRef) and draw that into the GraphicsContext displaylist,
which gets the image to the GPU process.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::snapshotAtSize):

Canonical link: https://commits.webkit.org/258399@main
